### PR TITLE
[SPARK-20841][SQL] Support table column aliases in FROM clause

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr added parsing rules to support table column aliases in FROM clause.

## How was this patch tested?
Added tests in `PlanParserSuite`,  `SQLQueryTestSuite`, and `PlanParserSuite`.

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/results/operators.sql.out [hunk 1]
```diff
@@ -1,5 +1,5 @@
 -- Automatically generated by SQLQueryTestSuite
--- Number of queries: 38
+-- Number of queries: 48
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/results/operators.sql.out]: This fix is not related to this pr though, this is not updated in https://github.com/apache/spark/commit/3c9eef35a85fd841efb4491b299345a2edf30e87

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 2]
```diff
@@ -711,7 +715,7 @@ nonReserved
     | ADD
     | OVER | PARTITION | RANGE | ROWS | PRECEDING | FOLLOWING | CURRENT | ROW | LAST | FIRST | AFTER
     | MAP | ARRAY | STRUCT
-    | LATERAL | WINDOW | REDUCE | TRANSFORM | USING | SERDE | SERDEPROPERTIES | RECORDREADER
```
Comment (gatorsmile) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Nit: keep it untouched.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/TableIdentifierParserSuite.scala [hunk 3]
```diff
@@ -49,7 +49,7 @@ class TableIdentifierParserSuite extends SparkFunSuite {
     "insert", "int", "into", "is", "lateral", "like", "local", "none", "null",
     "of", "order", "out", "outer", "partition", "percent", "procedure", "range", "reads", "revoke",
     "rollup", "row", "rows", "set", "smallint", "table", "timestamp", "to", "trigger",
-    "true", "truncate", "update", "user", "using", "values", "with", "regexp", "rlike",
```
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/TableIdentifierParserSuite.scala]: Nit: keep it unchanged.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 4]
```diff
@@ -472,13 +472,17 @@ identifierComment
     ;
 
 relationPrimary
-    : tableIdentifier sample? (AS? strictIdentifier)?      #tableName
+    : catalogTable                                         #tableName
```
Comment (gatorsmile) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Just change it to `tableIdentifier sample? tableAlias`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCDSQueryBenchmark.scala [hunk 5]
```diff
@@ -74,13 +74,13 @@ object TPCDSQueryBenchmark {
       // per-row processing time for those cases.
       val queryRelations = scala.collection.mutable.HashSet[String]()
       spark.sql(queryString).queryExecution.logical.map {
-        case ur @ UnresolvedRelation(t: TableIdentifier) =>
+        case ur @ UnresolvedRelation(t: TableIdentifier, _) =>
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCDSQueryBenchmark.scala]: Nit: `UnresolvedRelation(t: TableIdentifier, _)`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCDSQueryBenchmark.scala [hunk 6]
```diff
@@ -74,13 +74,13 @@ object TPCDSQueryBenchmark {
...
           queryRelations.add(t.table)
         case lp: LogicalPlan =>
           lp.expressions.foreach { _ foreach {
             case subquery: SubqueryExpression =>
               subquery.plan.foreach {
-                case ur @ UnresolvedRelation(t: TableIdentifier) =>
+                case ur @ UnresolvedRelation(t: TableIdentifier, _) =>
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCDSQueryBenchmark.scala]: Nit: `UnresolvedRelation(t: TableIdentifier, _)`

**Comment by gatorsmile** [comment]
Could you check the alias precedence of the other database?
```SQL
select col1 as a, col2 as b from t1 as (c, d);
```

Which alias should be used as the output schema? `(a, b)` or `(c, d)`?

**Comment by maropu** [comment]
ok, I'll check.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 2 above)
Comment (maropu) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: I found this `USING` entry caused a failure in `PlanParserSuite`;
https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala#L336
```
- joins *** FAILED ***
  == FAIL: Plans do not match ===
   'Project [*]                               'Project [*]
  !+- 'Join Inner                             +- 'Join UsingJoin(Inner,List(a, b))
      :- 'UnresolvedRelation `t`                 :- 'UnresolvedRelation `t`
  !   +- 'SubqueryAlias using                    +- 'UnresolvedRelation `u`
  !      +- 'UnresolvedRelation `u`, [a, b] (PlanTest.scala:97)
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/TableIdentifierParserSuite.scala (hunk 3 above)
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/TableIdentifierParserSuite.scala]: ditto

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 4 above)
Comment (maropu) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Ah, ok.

**Comment by maropu** [comment]
Currently, we have the same behaviour;
```
-- PostgreSQL
postgres=# create table t1(col1 int, col2 int);
CREATE TABLE

postgres=# select col1 as a, col2 as b from t1 as t(c, d);
ERROR:  column "col1" does not exist at character 8
STATEMENT:  select col1 as a, col2 as b from t1 as t(c, d);
ERROR:  column "col1" does not exist
LINE 1: select col1 as a, col2 as b from t1 as t(c, d);
               ^
-- MySQL
mysql> create table t1(col1 int, col2 int);
Query OK, 0 rows affected (0.01 sec)

mysql> select col1 as a, col2 as b from t1 as t(c, d);
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '(c, d)' at line 1
mysql> 


-- Spark with this pr
scala> Seq((1, 2), (2, 3)).toDF("col1", "col2").createOrReplaceTempView("t")

scala> sql("select col1 as a, col2 as b from t as t(c, d)").show
org.apache.spark.sql.AnalysisException: cannot resolve '`col1`' given input columns: [c, d]; line 1 pos 7;
'Project ['col1 AS a#18, 'col2 AS b#19]
+- SubqueryAlias t
   +- Project [col1#5 AS c#20, col2#6 AS d#21]
      +- SubqueryAlias t
         +- Project [_1#2 AS col1#5, _2#3 AS col2#6]
            +- LocalRelation [_1#2, _2#3]
```

**Comment by maropu** [comment]
@gatorsmile oh, you joined Databricks :))) congrats!

**Comment by gatorsmile** [comment]
It sounds like PostgresSQL supports it. See the docs in https://www.postgresql.org/docs/9.2/static/sql-select.html 

Actually, we also need to support the other alias in the from clauses:

See the link: https://drill.apache.org/docs/from-clause/ and http://docs.aws.amazon.com/redshift/latest/dg/r_FROM_clause30.html
```SQL
   with_subquery_table_name [ [ AS ] alias [ ( column_alias [, ...] ) ] ]
   table_name [ [ AS ] alias [ ( column_alias [, ...] ) ] ]
   ( subquery ) [ AS ] alias [ ( column_alias [, ...] ) ]
```

**Comment by maropu** [comment]
yea, I think so. I mean, in the name conflict case you described above, postgresql throws an error;
```
postgres=# select * from testTable;
 col1 | col2 
------+------
(0 rows)

postgres=# select * from testTable t(a, b);
 a | b 
---+---
(0 rows)

postgres=# select col1 a, col2 b from testTable t(c, d);
ERROR:  column "col1" does not exist at character 8
STATEMENT:  select col1 a, col2 b from testTable t(c, d);
ERROR:  column "col1" does not exist
LINE 1: select col1 a, col2 b from testTable t(c, d);
```
We donot support aliases for subquries. Should we include that support in this pr? Or, follow-up?

**Comment by maropu** [comment]
ping

**Comment by gatorsmile** [comment]
Yeah. That should be a negative case. The `FROM clause` should be resolved before `SELECT clause`.

The PR title is not accurate. I think we should keep the original JIRA name. `Support table column aliases in FROM clause`.

Yeah. These cases should be part of this JIRA. Please add the sub-tasks under this JIRA. Follow what Redshift documents and do them one by one?http://docs.aws.amazon.com/redshift/latest/dg/r_FROM_clause30.html 

```SQL
with_subquery_table_name [ [ AS ] alias [ ( column_alias [, ...] ) ] ]
table_name [ * ] [ [ AS ] alias [ ( column_alias [, ...] ) ] ]
( subquery ) [ AS ] alias [ ( column_alias [, ...] ) ]
table_reference [ NATURAL ] join_type table_reference
[ ON join_condition | USING ( join_column [, ...] ) ]
```

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: First, based on the link http://developer.mimer.se/validator/sql-reserved-words.tml, `USING` is a reserved word in SQL standard since SQL-92. 

Second, since 1.2, Hive introduces a flag `hive.support.sql11.reserved.keywords` for backward compatbility, which defaults to `true`. 
> Added In: Hive 1.2.0 with HIVE-6617: https://issues.apache.org/jira/browse/HIVE-6617
> Whether to enable support for SQL2011 reserved keywords. When enabled, will support (part of) SQL2011 reserved keywords.

In 2.2, Hive removes this flag and does not allow users to change it to false. That means, users are unable to use these reserved words as identifiers anymore, unless using them as quoted identifiers. See: https://issues.apache.org/jira/browse/HIVE-14872

Thus, I think it is safe to remove `USING` from the non-reserved words. 

cc @hvanhovell

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 7]
```diff
@@ -711,7 +711,7 @@ nonReserved
     | ADD
     | OVER | PARTITION | RANGE | ROWS | PRECEDING | FOLLOWING | CURRENT | ROW | LAST | FIRST | AFTER
     | MAP | ARRAY | STRUCT
-    | LATERAL | WINDOW | REDUCE | TRANSFORM | USING | SERDE | SERDEPROPERTIES | RECORDREADER
```
Comment (gatorsmile) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: First, based on the link http://developer.mimer.se/validator/sql-reserved-words.tml, `USING` is a reserved word in SQL standard since SQL-92. 

Second, since 1.2, Hive introduces a flag `hive.support.sql11.reserved.keywords` for backward compatbility, which defaults to `true`. 
> Added In: Hive 1.2.0 with HIVE-6617: https://issues.apache.org/jira/browse/HIVE-6617
> Whether to enable support for SQL2011 reserved keywords. When enabled, will support (part of) SQL2011 reserved keywords.

In 2.2, Hive removes this flag and does not allow users to change it to false. That means, users are unable to use these reserved words as identifiers anymore, unless using them as quoted identifiers. See: https://issues.apache.org/jira/browse/HIVE-14872

Thus, I think it is safe to remove `USING` from the non-reserved words. 

cc @hvanhovell

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 8]
```diff
@@ -592,7 +592,23 @@ class Analyzer(
     def resolveRelation(plan: LogicalPlan): LogicalPlan = plan match {
       case u: UnresolvedRelation if !isRunningDirectlyOnFiles(u.tableIdentifier) =>
         val defaultDatabase = AnalysisContext.get.defaultDatabase
-        val relation = lookupTableFromCatalog(u, defaultDatabase)
+        val foundRelation = lookupTableFromCatalog(u, defaultDatabase)
+
+        // If alias names assigned, add `Project` with the aliases
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Also gives a simple (SQL) example here to explain why we did this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 9]
```diff
@@ -592,7 +592,23 @@ class Analyzer(
...
+        val relation = if (u.outputNames.nonEmpty) {
+          val outputAttrs = foundRelation.output
+          // Checks if the number of the aliases is equal to expected one
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: > // Checks if the number of the aliases equals to the number of columns in the table.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 10]
```diff
@@ -592,7 +592,23 @@ class Analyzer(
...
+          if (u.outputNames.size != outputAttrs.size) {
+            u.failAnalysis(s"expected ${outputAttrs.size} columns but found " +
+              s"${u.outputNames.size} columns in alias names")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: >The number of column aliases does not match the number of column of the table `$tableName`. Number of column aliases: ${u.outputNames.size}; Number of columns: ${outputAttrs.size}.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 11]
```diff
@@ -36,8 +36,17 @@ class UnresolvedException[TreeType <: TreeNode[_]](tree: TreeType, function: Str
...
+ * We could add alias names for columns in a relation:
+ * {{{
+ *   // Assign alias names
+ *   SELECT col1, col2 FROM testData AS t(col1, col2);
+ * }}}
  */
-case class UnresolvedRelation(tableIdentifier: TableIdentifier) extends LeafNode {
+case class UnresolvedRelation(
+    tableIdentifier: TableIdentifier,
+    outputNames: Seq[String] = Seq.empty)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: Please add `@param` for these two parms

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: ok

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 7 above)
Comment (maropu) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: BTW, how did we decide these non-reserved words for Spark? It seems a lot of non-reserved words (e.g., `CUBE` and `GROUPING`) in Spark is the reserved ones in the ANSI standard...

**Comment by maropu** [comment]
@gatorsmile ok, thanks for your suggestion. I'll check the doc. and make sub-tasks there.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala [hunk 12]
```diff
@@ -131,8 +131,9 @@ object ResolveTableValuedFunctions extends Rule[LogicalPlan] {
         val outputAttrs = resolvedFunc.output
         // Checks if the number of the aliases is equal to expected one
         if (u.outputNames.size != outputAttrs.size) {
-          u.failAnalysis(s"expected ${outputAttrs.size} columns but " +
-            s"found ${u.outputNames.size} columns")
+          u.failAnalysis(s"Number of given aliases does not match number of output columns. " +
+            s"Function name: ${u.functionName}; number of aliases: " +
+            s"${u.outputNames.size}; number of output columns: ${outputAttrs.size}.")
         }
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala]: This fix is not related to this pr though, I modified along with this comment: https://github.com/apache/spark/pull/18079/files#diff-57b3d87be744b7d79a9beacf8e5e5eb2R604

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 13]
```diff
@@ -71,6 +84,11 @@ case class UnresolvedInlineTable(
  *   // Assign alias names
  *   select t.a from range(10) t(a);
  * }}}
+ * 
+ * @param functionName name of this table-value function
+ * @param functionArgs list of function arguments
+ * @param outputNames alias names of function output columns. If these names given, an analyzer
+ *                    adds [[Project]] to rename the output columns.
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: This fix is not related to this pr though, I modified along with this fix: https://github.com/apache/spark/pull/18079/files#diff-b4f9cbed8a042aeb12aeceb13b39d25aR50

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 7 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: I have added as much to the non-reserved keyword list as possible (without creating ambiguities). The reason for this is that many datasources (for instance twitter4j) unfortunately use reserved keywords for column names, and working with these was quite cumbersome. I took the pragmatic approach.

If we want to change this, then we need to do the same Hive did and create a config flag. We remove them for Spark 3.0...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 14]
```diff
@@ -676,9 +676,12 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
    * Create an aliased table reference. This is typically used in FROM clauses.
    */
   override def visitTableName(ctx: TableNameContext): LogicalPlan = withOrigin(ctx) {
-    val table = UnresolvedRelation(visitTableIdentifier(ctx.tableIdentifier))
-
-    val tableWithAlias = Option(ctx.strictIdentifier).map(_.getText) match {
+    val tableId = visitTableIdentifier(ctx.tableIdentifier)
+    val table = Option(ctx.tableAlias.identifierList) match {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Just if/else? Seems a bit heavy weight to wrap in an option...

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: ...or something like this:
```scala
val outputNames = Option(ctx.tableAlias.identifierList).map(visitIdentifierList).getOrElse(Nil)
val table = UnresolvedRelation(visitTableIdentifier(ctx.tableIdentifier), outputNames)
```

File: sql/core/src/test/resources/sql-tests/inputs/aliases.sql [hunk 15]
```diff
@@ -0,0 +1,17 @@
+-- Test data.
```
Comment (hvanhovell) [inline sql/core/src/test/resources/sql-tests/inputs/aliases.sql]: How about we name this file `table-aliases.sql`; that seems a little bit less confusing.

**Review by hvanhovell — APPROVED** [review]
LGTM from my end. I'll leave the final sign-off to @gatorsmile

File: sql/core/src/test/resources/sql-tests/inputs/aliases.sql (hunk 15 above)
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/aliases.sql]: ok

**Comment by gatorsmile** [comment]
LGTM too! : )

**Comment by gatorsmile** [comment]
Thanks! Merging to master.
